src/Val.cc

```diff
@@ -1301,24 +1302,60 @@ static std::variant<ValPtr, std::string> BuildVal(const rapidjson::Value& j, con
 
 			auto rt = t->AsRecordType();
 			auto rv = make_intrusive<RecordVal>(IntrusivePtr{NewRef{}, rt});
+
+			std::map<std::string, const rapidjson::Value*> normalized_keys;
+
+			// If key_func is given, map all JSON keys and store in above map.
+			if ( key_func )
+				{
+				for ( auto it = j.MemberBegin(); it != j.MemberEnd(); it++ )
+					{
+					ValPtr result;
+					try
+						{
+						result = key_func->Invoke(
+							zeek::make_intrusive<StringVal>(it->name.GetString()));
+						}
+					catch ( InterpreterException& )
+						{
+						/* Already reported. */
+						}
+
+					if ( ! result )
+						return "key function error";
+
+					normalized_keys[result->AsStringVal()->CheckString()] = &it->value;
+					}
+				}
+
+			// Now lookup record fields using the normalized input.
 			for ( int i = 0; i < rt->NumFields(); ++i )
 				{
-				auto td_i = rt->FieldDecl(i);
-				auto m_it = j.FindMember(td_i->id);
-				bool has_member = m_it != j.MemberEnd();
-				bool member_is_null = has_member && m_it->value.IsNull();
+				const auto td_i = rt->FieldDecl(i);
+				const rapidjson::Value* jval = nullptr;
+
+				if ( key_func )
+					{
+					auto m_it = normalized_keys.find(td_i->id);
+					jval = m_it != normalized_keys.end() ? m_it->second : nullptr;
+					}
+				else
+					{
+					auto m_it = j.FindMember(td_i->id);
+					jval = m_it != j.MemberEnd() ? &m_it->value : nullptr;
+					}
 
-				if ( ! has_member || member_is_null )
+				if ( ! jval || jval->IsNull() )
 					{
 					if ( ! td_i->GetAttr(detail::ATTR_OPTIONAL) &&
 					     ! td_i->GetAttr(detail::ATTR_DEFAULT) )
 						return util::fmt("required field %s$%s is %s in JSON", t->GetName().c_str(),
-						                 td_i->id, member_is_null ? "null" : "missing");
+						                 td_i->id, jval ? "null" : "missing");
```

Comment:
Previously the statement checked `jval && jval->IsNull()` via `member_is_null`, but now it checks if the jval itself is null. Is that intentional?

Reply:
Yes, but maybe this is a bit too clever... We only get here if `jval` is a nullptr (missing in the JSON object) or `jval->IsNull()` was true (it's a null JSON value). That's what `jval ? "null" : "missing"` uses.

Can bring back `has_member` and `member_is_null` bools, but they also seemed redundant at the time :-)

Reply:
Ah, yeah now that I look again the way you have it makes sense 👍🏼 Maybe add a comment there just for clarity.

Reply:
Done, are you otherwise fine with this PR?